core/src/test/scala/unit/kafka/log/LogTest.scala

```diff
@@ -116,6 +116,142 @@ class LogTest {
     assertHighWatermark(3L)
   }
 
+  private def assertNonEmptyFetch(log: Log, offset: Long, isolation: FetchIsolation): Unit = {
+    val readInfo = log.read(startOffset = offset,
+      maxLength = Int.MaxValue,
+      isolation = isolation,
+      minOneMessage = true)
+
+    assertFalse(readInfo.firstEntryIncomplete)
+    assertTrue(readInfo.records.sizeInBytes > 0)
+
+    val upperBoundOffset = isolation match {
+      case FetchLogEnd => log.logEndOffset
+      case FetchHighWatermark => log.highWatermark
+      case FetchTxnCommitted => log.lastStableOffset
+    }
+
+    for (record <- readInfo.records.records.asScala)
+      assertTrue(record.offset < upperBoundOffset)
+
+    assertEquals(offset, readInfo.fetchOffsetMetadata.messageOffset)
+    assertValidLogOffsetMetadata(log, readInfo.fetchOffsetMetadata)
```

Comment:
Is this useful since assertValidLogOffsetMetadata() calls the same log.read() for verification? Ditto in assertEmptyFetch().

Reply:
Yeah, good point. I think the main thing I wanted to assert here was that the segment position is defined and valid.